js/util/struct_array.js

```diff
@@ -0,0 +1,334 @@
+ * an array of typed structs. A StructArray is comprised of elements. Each element has a set of
+ * members that are defined when the `StructArrayType` is created.
+ *
+ * StructArrays useful for creating large arrays that:
+ * - can be transferred from workers as a Transferable object
+ * - can be copied cheaply
+ * - use less memory for lower-precision members
+ * - can be used as buffers in WebGL.
+ *
+ * @class StructArrayType
+ * @param {Array.<StructMember>}
+ * @param options
+ * @param {number} options.alignment Use `4` to align members to 4 byte boundaries. Default is 1.
+ *
+ * @example
+ *
+ * var PointArrayType = new StructArrayType({
+ *  members: [
+ *      { type: 'Int16', name: 'x' },
+ *      { type: 'Int16', name: 'y' }
+ *  ]});
+ *
+ *  var pointArray = new PointArrayType();
+ *  pointArray.emplaceBack(10, 15);
+ *  pointArray.emplaceBack(20, 35);
+ *
+ *  point = pointArray.get(0);
+ *  assert(point.x === 10);
+ *  assert(point.y === 15);
+ *  point._setIndex(1);
+ *  assert(point.x === 20);
+ *  assert(point.y === 35);
+ *
+ * @private
+ */
+function StructArrayType(options) {
+
+    var key = JSON.stringify(options);
+    if (structArrayTypeCache[key]) {
+        return structArrayTypeCache[key];
+    }
+
+    if (options.alignment === undefined) options.alignment = 1;
+
+    function StructType() {
+        Struct.apply(this, arguments);
+    }
+
+    StructType.prototype = Object.create(Struct.prototype);
+
+    var offset = 0;
+    var maxSize = 0;
+    var usedTypes = ['Uint8'];
+
+    StructType.prototype.members = options.members.map(function(member) {
+        member = {
+            name: member.name,
+            type: member.type,
+            components: member.components || 1
+        };
+
+        assert(member.name.length);
+        assert(member.type in viewTypes);
+
+        if (usedTypes.indexOf(member.type) < 0) usedTypes.push(member.type);
+
+        var typeSize = sizeOf(member.type);
+        maxSize = Math.max(maxSize, typeSize);
+        member.offset = offset = align(offset, Math.max(options.alignment, typeSize));
+
+        for (var c = 0; c < member.components; c++) {
+            Object.defineProperty(StructType.prototype, member.name + (member.components === 1 ? '' : c), {
+                get: createGetter(member, c),
+                set: createSetter(member, c)
+            });
+        }
+
+        offset += typeSize * member.components;
+
+        return member;
+    });
+
+    StructType.prototype.alignment = options.alignment;
+    StructType.prototype.BYTE_SIZE = align(offset, Math.max(maxSize, options.alignment));
+
+    function StructArrayType() {
+        StructArray.apply(this, arguments);
+    }
+
+    StructArrayType.serialize = serializeStructArrayType;
+
+    StructArrayType.prototype = Object.create(StructArray.prototype);
+    StructArrayType.prototype.StructType = StructType;
+    StructArrayType.prototype.BYTES_PER_ELEMENT = StructType.prototype.BYTE_SIZE;
```

Comment:
fix in 040ea50806a347221faf86691e3e33fd3acb7996

It was originally done to match https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray/BYTES_PER_ELEMENT but that's not a great idea because StructArray doesn't perfectly entire TypedArray interface and because StructArray types are defined on-the-fly.